### PR TITLE
GH-1893 / GH-1897: As a smith I want to refactor the builder to a snapshot-based IWorkspaceConfig handling

### PR DESCRIPTION
**Pull request by mor-n4**
Also includes:
GH-1897: As a smith I want to revisit the computation of workspace configuration updates

See #1893, #1897.

### CODE REVIEW

**Comment by mor-n4** [comment]
So far, this is the first step: replacing use of `ProjectDescription` by `XIProjectConfig` (as far as possible). Only remaining place where `ProjectDescription`  is still used is internally inside `ProjectBuilder`.

The next step will be to replace use of `XI(Workspace|Project)Config` by `(Workspace|Project)ConfigSnapshot` (as far as possible).

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/IOrderInfo.java [hunk 1]
```diff
@@ -52,6 +51,6 @@
 	public IOrderIterator<T> getIterator();
 
 	/** Creates a new instance of {@link IOrderIterator}. The given set of projects will be visited only. */
-	public IOrderIterator<T> getIterator(Collection<ProjectDescription> elements);
+	public IOrderIterator<T> getIterator(Collection<? extends T> elements);
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/IOrderInfo.java:54]: nice to make this generic, too

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/ProjectBuilder.java [hunk 2]
```diff
@@ -398,26 +400,39 @@ protected XBuildRequest newBuildRequest(
...
+		ProjectDescription.removeFromEmfObject(resourceSet);
+		ProjectDescription newPD = projectDescriptionFactory.getProjectDescription(projectConfig);
+		newPD.attachToEmfObject(resourceSet);
 	}
 
 	/** Create and configure a new resource set for this project. */
 	protected XtextResourceSet createNewResourceSet(ResourceDescriptionsData newProjectIndex) {
 		XtextResourceSet result = resourceSetProvider.get();
 		result.setURIResourceMap(uriResourceMap);
+		ProjectDescription projectDescription = projectDescriptionFactory.getProjectDescription(projectConfig);
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/ProjectBuilder.java:417]: Keeping this local (to the ProjectBuilder) looks much better!

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java [hunk 3]
```diff
@@ -246,16 +251,6 @@ public void clearResourceSets() {
 		}
 	}
 
-	/** @return all project descriptions. */
-	public List<ProjectDescription> getProjectDescriptions() {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java:250]: nice to have that removed!

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/lsp/N4JSWorkspaceConfig.java [hunk 4]
```diff
@@ -117,18 +114,6 @@ private WorkspaceChanges update(URI changedResource, Function<String, ProjectDes
 			update = update.merge(detectAddedRemovedProjects(oldProjects));
 		}
 
-		if (!update.getAddedProjects().isEmpty() || !update.getRemovedProjects().isEmpty()) {
-			// since the list of dependencies cached in class ProjectDescription does not contain names of projects that
-			// do not exist (in case of unresolved dependencies), we have to recompute all those lists of dependencies
-			// whenever a project is being added or removed:
-			for (IProjectConfig pc : oldProjects) {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/internal/lsp/N4JSWorkspaceConfig.java:124]: 👍

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/WorkspaceChanges.java [hunk 5]
```diff
@@ -260,7 +260,7 @@ public WorkspaceChanges merge(WorkspaceChanges changes) {
 				concat(addedSourceFolders, changes.addedSourceFolders),
 				newRemoved,
 				newAdded,
-				FluentIterable.concat(projectsWithChangedDependencies, projectsWithChangedDependencies)
+				FluentIterable.concat(projectsWithChangedDependencies, changes.projectsWithChangedDependencies)
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/WorkspaceChanges.java:263]: good find

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/XIWorkspaceConfig.java [hunk 6]
```diff
@@ -38,7 +36,7 @@
 	URI getPath();
 
 	/** Updates internal data based on changes of the given resource */
-	WorkspaceChanges update(List<URI> changedResources, Function<String, ProjectDescription> pdProvider);
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/XIWorkspaceConfig.java:41]: nice to get rid of this lambda

**Comment by mor-n4** [comment]
Builds green, merging ...

**Pull request merged**